### ISSUE TITLE
Compute gradient during the forward pass

### ISSUE THREAD

**Issue by AdrienCorenflos** [google/jax#3359]
Hi,

I have a specific use case where I can compute the gradient of my function efficiently during the forward pass. What is the preferred way of implementing the custom gradient in this case?

Thanks

What I have tried (reproducing a similar example without the real maths):

```python
import jax
from jax.lax import while_loop
import jax.numpy as jnp
from functools import partial


def example_loop_fwd(x, n, seed):
    return _example_loop_fwd(x, n, seed)


def _example_loop_fwd(x, n, seed):
    key = jax.random.PRNGKey(seed)
    grads = jnp.ones_like(x)
    def cond_fun(vars):
        _, _, _, i = vars
        return i < n

    def body_fun(vars):
        k, y, g, i = vars
        k, subkey = jax.random.split(k)
        uniforms = jax.random.uniform(subkey, shape=x.shape, minval=0.5, maxval=1.5)
        y = y * uniforms
        return key, y, g * uniforms, i + 1

    _, res, g, _ = while_loop(cond_fun, body_fun, (key, x, grads, 0))
    return jnp.sum(res), g


@partial(jax.custom_vjp, nondiff_argnums=(1, 2))
def example_loop(x, n, seed=0):
    res, g = _example_loop_fwd(x, n, seed)
    return res

def example_loop_bwd(n, seed, g, dres):
    return (g * dres,)


example_loop.defvjp(example_loop_fwd, example_loop_bwd)
```

**Comment by gnecula** [comment]
I would have done the same as you did. Does it work?

**Comment by mattjj** [comment]
Very cool! I think this is a case of "gradient preaccumulation" (Sec 10.2 of "Evaluating Derivatives"), where we can efficiently calculate the coefficients of a diagonal Jacobian during the forward pass and then just use them in a multiply in the backward pass. We used to have a helper function for this called `jarrett`, see e.g. #525 (the `_elementwise_std_basis` helper is directly analogous to the `jnp.ones_like(x)` call here), though it bitrotted so I deleted it at one point.

I probably would've written this as a custom_jvp, since then you'd get both forward- and reverse-mode support automatically:

```python
import jax
from jax.lax import fori_loop
import jax.numpy as jnp
from functools import partial


@partial(jax.custom_jvp, nondiff_argnums=(1, 2))
def example_loop(x, n, seed=0):
  keys = jax.random.split(jax.random.PRNGKey(0), n)

  def body_fun(i, y):
    uniforms = jax.random.uniform(keys[i], shape=y.shape)
    return y * uniforms

  y = fori_loop(0, n, body_fun, x)
  return jnp.sum(y)

@example_loop.defjvp
def example_loop_jvp(n, seed, primals, tangents):
  (x,), (x_dot,) = primals, tangents
  keys = jax.random.split(jax.random.PRNGKey(0), n)

  def body_fun(i, carry):
    y, jac = carry
    uniforms = jax.random.uniform(keys[i], shape=y.shape)
    return y * uniforms, jac * uniforms

  y, jac = fori_loop(0, n, body_fun, (x, jnp.ones_like(x)))
  return jnp.sum(y), jnp.sum(jac * x_dot)


print(example_loop(1., 5))  # 0.0006550594
print(jax.grad(partial(example_loop, n=5))(1.))  # 0.0006550594
```

I also wanted to show:
1. using `fori_loop` (or alternatively `scan`) for loops with a fixed trip count;
2. splitting the keys all at once outside the loop, which can be more performant.

In your real use case, it might be easier just to call `jax.jvp` on the primal `body_fun`, rather than writing a separate one for the primal and the jvp functions. It's just a little annoying with the loop counter (and potentially the PRNG key) in the way.

WDYT?

**Comment by mattjj** [comment]
Maybe we should revive the `jax.jarrett` function...

**Comment by AdrienCorenflos** [comment]
Thanks for all that Matt. Was a bit busy lately, I'll give it a try asap
